src/workerd/io/container-client.h

```diff
@@ -0,0 +1,88 @@
+#pragma once
+
+#include <workerd/rust/container/lib.rs.h>
+#include <workerd/rust/cxx-integration/cxx-bridge.h>
+
+#include <stdlib.h>
+
+#include <kj/async-io.h>
+#include <kj/debug.h>
+#include <kj/exception.h>
+#include <kj/mutex.h>
+
+#include <queue>
+
+namespace workerd::io {
+
+class ContainerStreamSharedState;
+
+// ContainerAsyncStream provides bidirectional communication with a container.
+//
+// This stream implements both reading and writing:
+// - Writing sends data to the container via the Rust service
+// - Reading receives messages from the container via MessageCallback queuing
+//
+// The stream uses shared state to coordinate between the MessageCallback
+// (which receives messages from the container asynchronously) and the
+// tryRead() method (which provides those messages to the C++ side).
+struct ContainerAsyncStream final: public kj::AsyncIoStream {
+  ContainerAsyncStream(::rust::Box<rust::container::ContainerService> service,
+      kj::Own<ContainerStreamSharedState> sharedState)
+      : service(kj::mv(service)),
+        sharedState(kj::mv(sharedState)) {}
+
+  void shutdownWrite() override {
+    service->shutdown_write();
+  }
+
+  kj::Promise<size_t> tryRead(void* buffer, size_t minBytes, size_t maxBytes) override;
+
+  kj::Promise<void> write(kj::ArrayPtr<const kj::byte> buffer) override {
+    if (service->write_data(buffer.as<Rust>())) {
+      return kj::READY_NOW;
+    } else {
+      return KJ_EXCEPTION(DISCONNECTED, "Write failed: stream is disconnected");
+    }
+  }
+
+  kj::Promise<void> write(kj::ArrayPtr<const kj::ArrayPtr<const kj::byte>> pieces) override {
+    for (auto piece: pieces) {
+      if (!service->write_data(piece.as<Rust>())) {
+        return KJ_EXCEPTION(DISCONNECTED, "Write failed: stream is disconnected");
+      }
+    }
+    return kj::READY_NOW;
+  }
+
+  kj::Promise<void> whenWriteDisconnected() override {
+    if (service->is_write_disconnected()) {
+      return kj::READY_NOW;
+    }
+    return kj::NEVER_DONE;
+  }
+
+ private:
+  ::rust::Box<rust::container::ContainerService> service;
+  kj::Own<ContainerStreamSharedState> sharedState;
+  kj::Maybe<kj::Array<kj::byte>> currentMessage;
+  size_t readOffset = 0;
+};
+
+// Shared state between MessageCallback and ContainerAsyncStream
+class ContainerStreamSharedState {
+ public:
+  ContainerStreamSharedState();
+
+  void enqueueMessage(::rust::Slice<const uint8_t> message);
+  kj::Maybe<kj::Array<kj::byte>> dequeueMessage();
+  kj::Promise<void> waitForMessage();
+
+ private:
+  kj::MutexGuarded<std::queue<kj::Array<kj::byte>>> messageQueue;
```

Comment:
is this shared?

Reply:
yes this is, it is shared between the rust tokio thread and the c++ worker thread.